### PR TITLE
[Beta Test] Wave Containers: add support for the Spack package manager

### PR DESCRIPTION
**Pull request by marcodelapierre**
This PR adds support for Spack as a software provider for Wave containers.
It comes out of some very fruitful conversations myself and Paolo have had recently.

It has the potential to pave the way for on-demand, architecture optimised containers with Nextflow and Wave! (some work yet to be completed). The key missing piece of work in this regard is the adoption of binary caches, to speed up build times (coming later).

Some points:

* Includes support for generalised architectures via `process.arch`
* Tested with a simple pipeline:
    * https://github.com/marcodelapierre/toy-cowsay-nf
* Unit tests written and successful
* Documentation updated
* @pditommaso has added back-end support for Spack YAML files

It is still a DRAFT for now, as 1 last key bit of work is missing:
- [ ] Use of templates for Spack Dockerfiles

@pditommaso @evanfloden have even more fun! :-)

@tgamblin @alalazo I hope this can be another useful contribution from the perspective of the Spack community, too!
@vsochat hopefully this will be merged soon enough .. watch this space ;)

### CODE REVIEW

**Comment by marcodelapierre** [comment]
@pditommaso again, not sure why I have some spurious commits from you.

I have just recloned the main nextflow repo and the merge says it is all good ...

**Comment by pditommaso** [comment]
What's wrong in your Git workflow? 😆 

screencapture-github-nextflow-io-nextflow-pull-3636-2023-02-10-16_22_51

**Comment by marcodelapierre** [comment]
> What's wrong in your Git workflow? 😆

Paolo I find it embarrassing 🤣🤣
I will dig further!

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (pditommaso) [inline plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:459]: What's the need for the `echo` ?

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (pditommaso) [inline plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:597]: Would not be better to use a template file instead of a big inline string?

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:459]: struggling to reproduce newlines appropriately in the Dockerfile, by using a single `echo` or `cat` ..

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:597]: good point, will be back on this one too

Comment (pditommaso) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:597]: Have a look a this class 

https://github.com/nextflow-io/nextflow/blob/762ba19c70006c84070bf1c736825e483c7fdde7/modules/nextflow/src/main/groovy/nextflow/executor/BashTemplateEngine.groovy#L31-L45

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy]: ok, addressed by using a single `echo` . templating (below) later on

**Comment by pditommaso** [comment]
This seems in very good shape. Roadmap to merge: 
1. move the Spack dockerfile to a template file 
2. avoid the use of exotic Bash option e.g. `-l` if possible, for consistency with other containers 
3. identify core `target` platforms to be supported 
4. add the support for Spack mirror on AWS for caching purpose 
5. implemented the corresponding support for Wave backend

**Comment by marcodelapierre** [comment]
> This seems in very good shape. Roadmap to merge:
> 
> 1. move the Spack dockerfile to a template file
> 2. avoid the use of exotic Bash option e.g. `-l` if possible, for consistency with other containers
> 3. identify core `target` platforms to be supported
> 4. add the support for Spack mirror on AWS for caching purpose
> 5. implemented the corresponding support for Wave backend

A memo note here, this is the github issue on arch compliance specifications by @pditommaso : https://github.com/nextflow-io/nextflow/issues/3701

**Comment by marcodelapierre** [comment]
OK, first successful run of a Wave Spack container using a `spack.yaml` file instead of a one-line package specification.

One issue to note. Every time I used Wave to build these Spack containers, the first time I build a new container the pull fails:
```
Command error:
  Unable to find image 'wave.seqera.io/wt/455e45c14efa/wave/build:4a71f62a456a4231cb0f0640b406628e' locally
  docker: Error response from daemon: received unexpected HTTP status: 502 Bad Gateway.
  See 'docker run --help'.
```
Then I just re-run the pipeline and everything goes OK.

Here I have some gaps in expertise. Not sure whether it is an issue arising somehow from Docker locally, or instead from your Wave backend. E.g. I am thinking at some latencies in making newly built layers available for pull....?

@pditommaso what do you think?

**Comment by pditommaso** [comment]
Have you got the build notification email? can't find anything wrong

**Comment by marcodelapierre** [comment]
The build was indeed successful (I have emailed you the notification), as I was suspecting.
So it is something that happens afterwards.

**Comment by pditommaso** [comment]
Umm, `502 Bad Gateway` sounds more like a load balancer glitch. I can pull the container. Have you tried to build it again?

**Comment by marcodelapierre** [comment]
So, Paolo, I was able to run some tests with both `spack` and `conda` Wave builds.

I confirm that once the image is built by the backend, any pull after the first (failing, if applicable) one will work.

Is there a way to force the backend to rebuild the image.
What I have done in my tests is to change a package version to force a new build.

So far: 
* all builds succeeded
* all Spack Wave builds failed on *first pull*
* I have identified at least one large Conda Wave build that fails on *first pull*, too: `process.conda = 'conda-forge::cowpy https://conda.deepmodeling.com::deepmd-kit=*=*cpu https://conda.deepmodeling.com::libdeepmd=*=*cpu'` (3 GB image, DeepMD is big)

Notably, the error I got with Conda was slightly different:
```
Command error:
  Unable to find image 'wave.seqera.io/wt/c7e9f7bb170d/wave/build:593503d65400503e25f885163ecdc304' locally
  docker: Error response from daemon: missing or empty Content-Length header.
  See 'docker run --help'.
```

So it really seems to me about something going badly on the backend side between the build and the client request to pull.

Thoughts?

**Comment by pditommaso** [comment]
Can I use this fork to replicate the error?

**Comment by marcodelapierre** [comment]
Yes, but I imagine (not tested) that you can replicate the `conda` case with the develop branch as well.

**Comment by pditommaso** [comment]
I'd like to focus on the spack problem because it looks more consistent. What is the config/script you are using ?

**Comment by marcodelapierre** [comment]
Use this Paolo:
```
git clone git@github.com:marcodelapierre/toy-cowsay-nf 
cd toy-cowsay-nf 
nextflow run main.nf -profile spackwaveyaml
```
All of the required files are in that github repo, in particular `nextflow.config` and `spack.yaml`.

Right now, you can only do multi-packages (e.g. `samtools blast`) with the YAML file, due to a silly missing bit of implementation for the other case.

The corresponding container image should already be in the Wave cache as I built it yesterday, so I suggest you open the `spack.yaml` and change `nano ^ncurses abi=5` to for instance `nano@6.0 ^ncurses abi=5` (to get always new images you can use a variety of Nano versions down to `4.0`).

(Note nano is not needed by the pipeline, it is really just a way to add some extra minimal complexity to the image build, for testing purposes)

**Comment by marcodelapierre** [comment]
Ah! the moment I read the error  message, I realised what is the issue. Yesterday I had changed the codebase to use `spack.yml` instead of `spack.yaml` .. well Spack does not like it!

I have fixed the code, you can go ahead now with the testing, apologies for the hiccup!

To help us remembering about this, I recommend to also tweak the backend to use `spack.yaml`; let me know when you do it and I will change a couple of corresponding lines in the client code.

File: NOTICE
Comment (pditommaso) [inline NOTICE:147]: Not sure this should be added. What code is taken from Spack?

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (pditommaso) [inline plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:502]: Worth to move this into a template file

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (pditommaso) [inline plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:617]: also this it's better to keep into a template

**Comment by pditommaso** [comment]
> To help us remembering about this, I recommend to also tweak the backend to use spack.yaml; let me know when you do it and I will change a couple of corresponding lines in the client code.

Ok, changed this

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:502]: agree, it's in my TODO

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:617]: agree, it's in my TODO

File: NOTICE
Comment (marcodelapierre) [reply NOTICE:147]: So, I added this because I used their Dockerfile template as a starting point for ours:
https://github.com/spack/spack/blob/develop/share/spack/templates/container/Dockerfile

Not sure whether this is the best way to acknowledge it within Nextflow, what do you reckon?

**Comment by marcodelapierre** [comment]
> > To help us remembering about this, I recommend to also tweak the backend to use spack.yaml; let me know when you do it and I will change a couple of corresponding lines in the client code.
> 
> Ok, I've changed this

Thanks, I have just updated the client.

**Comment by marcodelapierre** [comment]
Hi @pditommaso, plan update on this one:
- will restart on this early next week;
- will merge master with #3786 (generalised `arch` implementation) as soon as approved;
- currently targeting Tue 2 May as worst-case date to finalise this PR.

**Comment by pditommaso** [comment]
I had missed this, it looks a great plan 👍

**Comment by marcodelapierre** [comment]
So @pditommaso , on this one:
> avoid the use of exotic Bash option e.g. `-l` if possible, for consistency with other containers

Right now I have this in the Dockerfile:
```
ENTRYPOINT ["/bin/bash", "--rcfile", "/etc/profile", "-l", "-c", "\$*", "--" ]
CMD [ "/bin/bash" ]
```

The reason for this  is:
```
spack env activate --sh -d . >> /etc/profile.d/z10_spack_environment.sh
```

That is a set of variable configurations to enable the installed Spack packages to execute correctly, things like PATH and so on.

Unfortunately it does not seem trivial nor desirable to generalise it in a hard-codeable way in the Dockerfile. The reason is that some of those variable definitions are dependent on the specific packages being installed. E.g. each Python package installed in the container needs its own specific `PYTHONPATH`.

As such, I would live the ENTRYPOINT as is at this stage.
In the future, a chat with the Spack devs might be useful in figuring out more optimal ways to handle these aspects at the stage of the Spack deployment of packages such as Python ones.

What do you think?

**Comment by pditommaso** [comment]
How that is different from Micromamba ?

https://github.com/mamba-org/micromamba-docker/blob/main/Dockerfile#L60

**Comment by marcodelapierre** [comment]
ah sure, if you are happy with using a service script, I will do that way, thanks!

**Comment by marcodelapierre** [comment]
OK, updated the ENTRYPOINTs in the Dockerfile templates

**Comment by pditommaso** [comment]
Cool! does it mean this PR is ready for review?

**Comment by marcodelapierre** [comment]
Two remaining open points:
* integration with `process.arch`
* refactoring of the Dockerfile templates

At this stage, I am on track to have the PR ready for review by next Monday.

**Comment by pditommaso** [comment]
Update your local copy

**Comment by marcodelapierre** [comment]
> Update your local copy

Done, thanks, I noticed the docs overhaul, going from `rst` to `md`

**Comment by marcodelapierre** [comment]
@pditommaso  for your Docker platform spec (x86, arm, ..), is the backend using the `containerPlatform` string, as sent by the Wave plugin via `SubmitContainerTokenRequest` ?
```
        return new SubmitContainerTokenRequest(
                containerImage: assets.containerImage,
                containerPlatform: assets.containerPlatform,
                containerConfig: containerConfig,
                containerFile: assets.dockerFileEncoded(),
                condaFile: assets.condaFileEncoded(),
                spackFile: assets.spackFileEncoded(),
                buildRepository: config().buildRepository(),
                cacheRepository: config.cacheRepository(),
                timestamp: OffsetDateTime.now().toString(),
                fingerprint: assets.fingerprint()
        )
```

**Comment by pditommaso** [comment]
Nope uses `ContainerPlatform` to handle it 

```
   @Unroll
    def 'should validate platform' () {
        expect:
        ContainerPlatform.of(PLATFORM) == EXPECTED
        ContainerPlatform.of(PLATFORM).toString() == STRING

        where:
        PLATFORM        | EXPECTED                                      | STRING
        null            | new ContainerPlatform('linux','amd64')        | 'linux/amd64'
        'x86_64'        | new ContainerPlatform('linux','amd64')        | 'linux/amd64'
        'amd64'         | new ContainerPlatform('linux','amd64')        | 'linux/amd64'
        'linux/amd64'   | new ContainerPlatform('linux','amd64')        | 'linux/amd64'
        and:
        'arm'           | new ContainerPlatform('linux','arm','v7')     | 'linux/arm/v7'
        'linux/arm'     | new ContainerPlatform('linux','arm','v7')     | 'linux/arm/v7'
        'linux/arm/7'   | new ContainerPlatform('linux','arm','v7')     | 'linux/arm/v7'
        'linux/arm/v7'  | new ContainerPlatform('linux','arm','v7')     | 'linux/arm/v7'
        'linux/arm/5'   | new ContainerPlatform('linux','arm','v5')     | 'linux/arm/v5'
        'linux/arm/v5'  | new ContainerPlatform('linux','arm','v5')     | 'linux/arm/v5'
        and:
        'arm64'         | new ContainerPlatform('linux','arm64')        | 'linux/arm64'
        'linux/arm64'   | new ContainerPlatform('linux','arm64')        | 'linux/arm64'
        'linux/aarch64' | new ContainerPlatform('linux','arm64')        | 'linux/arm64'
        and:
        'linux/arm64/v8'| new ContainerPlatform('linux','arm64')        | 'linux/arm64'
        'linux/arm64/v7'| new ContainerPlatform('linux','arm64','v7')   | 'linux/arm64/v7'

    }
```

**Comment by marcodelapierre** [comment]
yep, but where do you get `PLATFORM` from?  some info expected from the client right?

**Comment by pditommaso** [comment]
It's the value coming from the [request](https://github.com/nextflow-io/nextflow/pull/3636#issuecomment-1527198694) above!

**Comment by marcodelapierre** [comment]
ok, perfect, thanks

**Comment by marcodelapierre** [comment]
Hi @pditommaso , some updates and some questions for you:

1. `process.arch` implemented in Wave+Spack !
2. only remaining minimal bit (under the hood and localised) is using more appropriate templates for the 2 Spack dockerfiles
3. there are two issues that I think require your attention
4. as such, you can start and review the PR, as the only open point on my side is 2. above

Now for the two hot issues for you:
1. the gateway/container pull error that happens every first time that Waves create a new Spack container; we were saying above that this might come from some in-balance in the backend. I imagine it may relate to the big size of some layers in the Spack Wave build. As a partial proof, a few days ago I was testing small changes at the end of the Dockefile (ENTRYPOINT), so new containers were reusing cached layers for the package installation steps, and were NOT failing to pull. To reproduce the error this should be enough `nextflow run marcodelapierre/toy-cowsay-nf -profile spackwave` (define the variable TOWER_ACCESS_TOKEN)
2. issue in `WaveClient.groovy` with the method `defaultFusionUrl()`. This was using the `containerPlatform` property, that was defined at the Wave level. However, the platform variable is now defined at the process level. I have got stuck when trying to refactor in this direction; that method is used by `resolveContainerConfig()`, which among others is used by `sendRequest(String image)`, for which I am struggling to figure out how to pass the process specific platform. As a result, right now some tests within `WaveClientTest.groovy` are inconsistent and/or output errors. Any tip/help here would be much appreciated!

File: NOTICE
Comment (pditommaso) [reply NOTICE:146]: Then, the apache header should go also in the dockerfile

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/config/WaveConfig.groovy
Comment (pditommaso) [inline plugins/nf-wave/src/main/io/seqera/wave/plugin/config/WaveConfig.groovy:76]: It'd be better to turn this into a constant and use it also in the check below (line 88)

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (pditommaso) [inline plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:239]: I think this can be solved by passing the container platform as parameter to `defaultFusionUrl` and `resolveContainerConfig`. 

When the platform is not available, like in the context of `sendRequest(String image)`, you can yse the default container platform.

**Comment by pditommaso** [comment]
Great work Marco! Regarding the open points: 

1. I'm start to think it can be a problem with the wave backend using a spot instance for the build. It may be not available the very first time.
2. I've commented in the corresponding method 
3. is dropping the support for `ARM32` an option? I don't think it's really used

File: NOTICE
Comment (marcodelapierre) [reply NOTICE:146]: done!

File: modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy:92]: minor edit done.

for the unit test, look at this one: 

https://github.com/marcodelapierre/nextflow/blob/add/wave-spack/modules/nextflow/src/test/groovy/nextflow/processor/ArchitectureTest.groovy#L38-L52

Here, column `DOCK` tests the routine `validateArchToDockerArch()`, which you mention, whereas column `SPACK` tests `validateArchToSpackArch()`.

Do you think an additional test would help?

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:238]: ahah good old coding school at non-CS uni classes LOL

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:336]: done!

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:367]: Good point! However, I added this one by blindly copying the `conda` case above, i.e. `L347-L348` match `L366-L367`. I haven't tested the following,but I actually think these checks should be taken off, because `attrs` has already been validated above at `L327-L330`:
```
        if( config().strategy() )
            attrs = resolveConflicts(attrs, config().strategy())
        else
            checkConflicts(attrs, task.lazyName())
```

I would let you approve whether to delete or not.

**Comment by marcodelapierre** [comment]
> is dropping the support for ARM32 an option? I don't think it's really used. Alternatively, when  ARM32 is specified, it could be thrown an exception in the defaultFusionUrl method saying Fusion does not support ARM32 architecture.

Happy to drop it, unless you know of user cases in commercial cloud.

**Comment by pditommaso** [comment]
> Happy to drop it, unless you know of user cases in commercial cloud

Not really!

File: modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy:92]: I see. It's tested indirectly via the constructor. Fair enough, then better to declare it `private`

**Comment by marcodelapierre** [comment]
> > Happy to drop it, unless you know of user cases in commercial cloud
> 
> Not really!

done!

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (pditommaso) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:367]: melius abundare .. :D ok, let's just add `Unexpected spack and dockerfile conflict while resolving wave container`, same for conda

File: modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy:92]: done!

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:367]: agree! done!

**Comment by marcodelapierre** [comment]
> I'm start to think it can be a problem with the wave backend using a spot instance for the build. It may be not available the very first time.

Just a memo that It is available when the newly built layers are small

**Comment by pditommaso** [comment]
Could not be a timeout problem?

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/config/WaveConfig.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/config/WaveConfig.groovy:76]: done!

**Comment by marcodelapierre** [comment]
> Could not be a timeout problem?

the container build succeeds, in fact I get notified via email. It is the subsequent pull that fails.

**Comment by pditommaso** [comment]
Weird. I'll give a try once you have done

**Comment by pditommaso** [comment]
Just checked, and it worked on first try 🤷‍♂️

**Comment by marcodelapierre** [comment]
> Just checked, and it worked on first try 🤷‍♂️

Did you use my pipeline with `-profile spackwave`? I think it worked because the default image is still cached in Wave.

I think you can play with the Dockerfile, `WaveClient.groovy` line 579 and onwards.
First, you add a directive like `RUN echo "hello" >/world` at the end of the Dockerfile, so new layer at the end , should be fine.
But then you add it at the very beginning, invalidating the cache.

**Comment by pditommaso** [comment]
>. Did you use my pipeline with -profile spackwave

yes. let me try again clearing the cache

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:239]: so, just to confirm, in `WaveRunCmd.groovy` the `runContainer()` method does not need to know about Fusion?

**Comment by pditommaso** [comment]
Is this the error? 

```
  Unable to find image 'wave.stage-tower.net/wt/2e46a336f730/wave/build/stage:542bc2ddaeecfc5156ab07c0e1f6fcb2' locally
  docker: Error response from daemon: manifest for wave.stage-tower.net/wt/2e46a336f730/wave/build/stage:542bc2ddaeecfc5156ab07c0e1f6fcb2 not found: manifest unknown: repository '195996028523.dkr.ecr.eu-west-1.amazonaws.com/wave/build/stage:542bc2ddaeecfc5156ab07c0e1f6fcb2' not found.
  See 'docker run --help'.


```

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (pditommaso) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:239]: Not directly. Maybe the best thing is to add a optional `platform` parameter that default to [DEFAULT_DOCKER_PLATFORM](https://github.com/nextflow-io/nextflow/blob/main/plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy#L65-L65)

**Comment by marcodelapierre** [comment]
> Is this the error?
> 
> ```
>   Unable to find image 'wave.stage-tower.net/wt/2e46a336f730/wave/build/stage:542bc2ddaeecfc5156ab07c0e1f6fcb2' locally
>   docker: Error response from daemon: manifest for wave.stage-tower.net/wt/2e46a336f730/wave/build/stage:542bc2ddaeecfc5156ab07c0e1f6fcb2 not found: manifest unknown: repository '195996028523.dkr.ecr.eu-west-1.amazonaws.com/wave/build/stage:542bc2ddaeecfc5156ab07c0e1f6fcb2' not found.
>   See 'docker run --help'.
> ```

No, I typically get one of these two:
```
Command error:
  Unable to find image 'wave.seqera.io/wt/455e45c14efa/wave/build:4a71f62a456a4231cb0f0640b406628e' locally
  docker: Error response from daemon: received unexpected HTTP status: 502 Bad Gateway.
```

```
Command error:
  Unable to find image 'wave.seqera.io/wt/c7e9f7bb170d/wave/build:593503d65400503e25f885163ecdc304' locally
  docker: Error response from daemon: missing or empty Content-Length header.
  See 'docker run --help'.
```

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:239]: now updating the unit tests

Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:239]: (but having dinner soon)

**Comment by pditommaso** [comment]
Weird, cannot even find it in the logs. When it happen again let me know

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy:239]: 😂

Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy]: Duff always helps .. done!

**Comment by marcodelapierre** [comment]
Ok, so:
- implemented all your feedback from the past few days
- sorted out issue with `defaultFusionUrl`
- dropped `ARM32`

Tomorrow I will refactor the Dockerfile templates, and then release this PR for your final review and merging.

**Comment by pditommaso** [comment]
Awesome!

**Comment by marcodelapierre** [comment]
@pditommaso I am realising the PR for review and merge!

There are TWO outstanding issues, the first one being new:

1. I have moved to Dockerfile templates, the code builds and the unit tests pass, but .. at runtime I get this error:
```
ERROR ~ Error executing process > 'sayHello (2)'

Caused by:
  Cannot invoke "java.net.URL.openConnection()" because "url" is null
```
e.g. with this trace:
```
java.lang.NullPointerException: Cannot invoke "java.net.URL.openConnection()" because "url" is null
        at org.codehaus.groovy.runtime.ResourceGroovyMethods.configuredInputStream(ResourceGroovyMethods.java:2199)
        at org.codehaus.groovy.runtime.ResourceGroovyMethods.newReader(ResourceGroovyMethods.java:2271)
        at io.seqera.wave.plugin.WaveClient.spackRecipeToDockerFile(WaveClient.groovy:528)
```
it must be something very silly that I am missing about the usage of these methods ... tips?


2. the backend pull errors .. I will run a few other tests once the above is fixed. The dockerfile template makes it much quicker to prototype variations of the dockerfile, that can help troubleshooting this one, e.g. by altering the creation and caching of layers (number and size)

**Comment by pditommaso** [comment]
> have moved to Dockerfile templates, the code builds and the unit tests pass, but .. at runtime I get this error

I believe is happening because BashWrapperBuilder is in another module, try using 

```
        final template = WaveClient.class.getResource('/dockerfile-spack-recipe.txt')
        try(final reader = template.newReader()) {
            final result = new BashTemplateEngine().render(reader, binding)
            return result
        }
```

The key is `WaveClient.class.getResource`, if it works it would also be better to move the files under the path `templates/spack`



2. I'll look into once we merge this

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy]: done

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy]: done

File: plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy
Comment (marcodelapierre) [reply plugins/nf-wave/src/main/io/seqera/wave/plugin/WaveClient.groovy]: done

**Comment by marcodelapierre** [comment]
So the last man standing is:

> the backend pull errors .. I will run a few other tests once the above is fixed. The dockerfile template makes it much quicker to prototype variations of the dockerfile, that can help troubleshooting this one, e.g. by altering the creation and caching of layers (number and size)

**Review by pditommaso — APPROVED** [review]
Thanks Marco! awesome job!

**Comment by marcodelapierre** [comment]
Thank you Paolo, let's talk at the next PR...!

**Comment by marcodelapierre** [comment]
@tgamblin @alalazo @vsoch it is merged! :-) 

the final bit will be setting up a binary cache of bioinformatics packages to render the service useful...I might ping you on this after my conference frenzy in May!   Todd, Max  see you at ISC!

**Pull request merged**
